### ISSUE TITLE
Add a component that allows to run recording rules on Grafana Agent

### ISSUE THREAD

**Issue by mikemykhaylov** [grafana/alloy#414]
### Background

In our K8s environments, one of the types of metrics we collect with Grafana Agent are the network flows between pods and outside IPs. Due to the clusters' size, as well as having explicit "source" and "destination" labels on the metrics, the cardinality is exploding, easily reaching above 10k metrics reported per node. We would like to conditionally reduce the cardinality of a couple of metrics, preserving specific labels and dropping others, prior to doing a remote write to Mimir. 

Using `prometheus.relabel` to drop labels is not an option, because in our use case, it would generate duplicate metrics (i.e. there are metrics that are the same by all labels except those that we would like to drop), and so Prometheus would drop them ([link to post explaining this](https://grafana.com/blog/2022/10/20/how-to-manage-high-cardinality-metrics-in-prometheus-and-kubernetes/)). What we need instead is, paraphrasing the article
```
my_metric_total{env=“dev”, ip=“1.1.1.1"} 1
my_metric_total{env=“dev”, ip=“3.3.3.3"} 3
my_metric_total{env=“dev”, ip=“5.5.5.5"} 7

# Dropped label ip
my_metric_total{env=“dev”} 11
```
Assuming the multitude of labels that would be tedious to enumerate, an obvious solution to this would be the use of recording rules. Namely, rule like the following would work (followed by the relabel dropping the original metrics)
```yaml
rules:
  - record: my_metric_total_noip
    expr: sum without (ip) (my_metric_total)
```
However, if we set up this rule in Mimir, this doesn't really help, as the metrics are already stored. If, instead, the rules ran on the Agent, only the Agent would have to deal with the metrics, while Mimir would only see the reduced-cardinality metrics

### Proposal

We add a new component to the Flow mode of Grafana Agent along the lines of `prometheus.recordrules` that allows us to specify recording rules in a similar way to Prometheus. We could have a thin `storage.Appendable` and `storage.Queryable` implementation which is exported as a receiver and appended to; we would then convert our rule configuration into Prometheus-native `RuleGroup`, evaluate it, and register new metrics in the next component's receiver.

**Comment by wildum** [comment]
Hey, thanks for opening this proposal, I think that it makes a lot of sense :) We have a new [proposal review process](https://github.com/grafana/alloy/blob/main/docs/design/908-proposal-process.md). I'd be happy to move your proposal to the active stage if you could add a little bit more details (could you rework it to use our [template](https://github.com/grafana/alloy/blob/main/docs/design/template.md)?) What do you think @mikemykhaylov?

**Comment by wildum** [comment]
A very similar idea is being implemented in [prometheus](https://github.com/prometheus/prometheus/pull/10529). If it's accepted we could integrate it in the scrape component of Alloy

**Comment by mikemykhaylov** [comment]
Sure, I'd be glad to rework it to the template! The Prometheus PR does look like something that could be used here; if that work wraps up I'll definitely take a crack at adding it to Alloy, in part as an excuse to get a refresher on Go 😅

**Comment by wildum** [comment]
I'm moving this proposal to "Active" state to discuss it more in-depth with the team and reach a consensus.

**Comment by bingkunyangvungle** [comment]
Just to follow up, do we have some updates on this? Adding a component that support it would be super helpful to reduce the cardinality.

**Comment by wildum** [comment]
Hey,

One concern about it was expressed in the [Prometheus PR](https://github.com/prometheus/prometheus/pull/10529): How should counter resets be handled?
For example, if I aggregate two counters metrics A and B, if B resets but A doesn't, then my aggregated metric will be wrong. 
If I understand correctly, this is only an edge case because counters should usually reset at the same time.
One solution to handle proper counter reset would be to do something similar to what VictorMetrics is doing: https://docs.victoriametrics.com/stream-aggregation/#total_prometheus.
For the first version, I would suggest documenting this issue and seeing whether a proper handling like this is worth it.

**Unless there are blocking concerns to the proposal, I will move it to "Likely accepted" by the end of the week. Then we will wait one more week to wait in case a major concern arises. If not then we will move the proposal to "Accepted" and anyone will be free to implement it (please mention it in this issue before starting the work to avoid having several people doing it in parallel)**

**Comment by wildum** [comment]
The proposal is accepted :) As stated above, anyone is free to pick it up, just mention it before starting the implementation to avoid duplication

**Comment by thampiotr** [comment]
How should this work with clustering? For example:

```
my_metric_total{env=“dev”, ip=“1.1.1.1"} 1      # ends up on instance A
my_metric_total{env=“dev”, ip=“3.3.3.3"} 3.  # ends up on instance B
my_metric_total{env=“dev”, ip=“5.5.5.5"} 7.   # ends up on instance A
```
The metrics required to aggregate would be on different Alloy instances, since we shard the targets. How would we want to handle this? 
Or is this component meant to run only on single-instance deployments where scraping is not distributed?

**Comment by sjentzsch** [comment]
Same as Prometheus handles it when sharding is activated.
From my perspective, all recording rules should be evaluated on all shards. You would not have any overlap, as the metric base does not overlap. Only the scrape jobs are sharded, not the recording rules.

**Comment by mattdurham** [comment]
Thinking on this more this  would not be a direct port/conversion of prometheus but something a bit different. Conceptually more akin to adaptive metrics. I think @sjentzsch is that it only applies to a single target scrape, the actual proposal might be closer to something like adapative metrics.

**Comment by thampiotr** [comment]
> Same as Prometheus handles it when sharding is activated. 

What are you referring to when you say Prometheus sharding? I'm not sure [which Prometheus feature](https://www.google.com/search?q=site%3Aprometheus.io%2Fdocs+sharding) you're referring to. 

> From my perspective, all recording rules should be evaluated on all shards. You would not have any overlap, as the metric base does not overlap. Only the scrape jobs are sharded, not the recording rules.

If we did the above on my example, let's say, to calculate the `avg without(ip) (...)`:

```
my_metric_total{env=“dev”, ip=“1.1.1.1"} 1      # ends up on instance A
my_metric_total{env=“dev”, ip=“3.3.3.3"} 3.  # ends up on instance B
my_metric_total{env=“dev”, ip=“5.5.5.5"} 7.   # ends up on instance A
```

We would get something like this:

```
my_metric_total{env=“dev”} 4      # calculated on instance A
my_metric_total{env=“dev”} 3     # calculated on instance B
```

without any further handling, one of these writes would fail with duplicate sample / out-of-order.

**Comment by thoro** [comment]
I was just looking for such a feature, specifically for the node_cpu_seconds_total of node_exporter, it's creating thousands of metrics I'm really not interested, and an aggregation at scrape time would be great (i.e. drop cpu label, and sum on mode)

@thampiotr I don't see how your example would work out in the real world. Sharding does not switch the scrape targets around, so one instance will still scrape the same target. 
Additionally the sum would be on a per scrape basis, and not over multiple scrapes. If the metric is provided by multiple instances of the application, there would be additional labels, which can't be aggregated anyways.

**The issue that one metric ends up on a different instance can't happen.**

**Comment by oriInsg** [comment]
+1 from me for the exact use case and metric described by @thoro above.

**Comment by Liquidmasl** [comment]
Might be related?
I would need to be able to get some min, avg, max of metrics per minute and log that, so I dont have to store to many single samples on the metrics server. 
this could be done with prometheus rules. 
Hopefully this will be possible here as well

**Comment by thampiotr** [comment]
> The issue that one metric ends up on a different instance can't happen.

@thoro It can if you are aggregating across targets. The aggregation within one target is possible, but that's an important limitation that we'd need to call out. I don't really mind if we did aggregation within one target only since it seems it addresses most of people's use cases here?

**Comment by thoro** [comment]
Yes, I was only thinking about single target aggregation

**Comment by ravishankar15** [comment]
I was checking the code for implementing this feature, 
1. I understand the part of prometheus adding the support to their manager as we where discussing in the above which could be a easier integration in alloy
2. There was also a discussion to have a separate component which would aggregate all the data from the single scrape.

for (2) I think we should have a **separate frequency** with WAL so that i flush the aggregated data to the next component something similar to remote_write.


The separate frequency can be different from the scrape I believe. I feel its okay shall i go ahead and implement the (2) ?

**Comment by thampiotr** [comment]
@ravishankar15 wouldn't it be possible to implement a new component using a Prometheus interceptor? Similar to prometheus.enrich component. I'm not sure if WAL is needed or desired here. 

Another option could be to have a component that acts as a proxy for scraping - so you would scrape that new component with prometheus.scrape, which would go to the actual target, do the aggregation and then produce an aggregated scrape response. 

Given a few options here, it may be good idea to discuss pros/cons and design the implementation.

**Comment by ravishankar15** [comment]
@thampiotr The problem with the interceptor is that it gets called with each entry and query like `sum` will not take effect I believe.

Yeah a separate component something like, the new component say `prometheus.recordrules` will collect the scrape entries for a defined interval like 10s(this can be different from the scrape time interval) or so and then runs the aggregation on the data from that 10s window and pass it down to the next component. We cannot use the interceptors directly for this approach we might need to buffer the values in the component and then trigger the interceptor on the aggregated value.

I was reading the threads but could not find a conclusion on which approach we are aligned.

**Comment by thampiotr** [comment]
> I was reading the threads but could not find a conclusion on which approach we are aligned.

There's no alignment yet, I think we probably want to reach some agreement before implementing this. 

I see, the appender is receiving one sample at a time, so this may be more challenging. But at scrape time you get all the metrics exposed by the target in one go, which seems perfect for aggregation. 


The flow of data in this solution could look like this:
```
actual_target <--scrape-- prometheus.aggregate_target <-- scrape--- prometheus.scrape --write-samples--> prometheus.remote_write
```

And the promethus.aggregate_target would act very similar to a Prometheus exporter actually. Perhaps we can even split this out into an exporter project and make it an embedded `prometheus.exporter.aggregator` component 🤔

**Comment by ravishankar15** [comment]
@thampiotr Can you check this out ? https://github.com/prometheus/proposals/pull/67/files looks like we might be able to see this in prometheus soon. I will keep a tap on that proposal incase not picked will try to make the change in prometheus and integrate the same in alloy. What do you think ?

**Comment by thampiotr** [comment]
Thanks! We typically prefer upstream-first solutions, so would be very much in favour of making it happen in Prometheus. Will upvote.
